sources/src/parser/cityobjectelementparser.cpp

```diff
@@ -35,88 +38,100 @@ namespace citygml {
+                typeIDTypeMap.insert(HANDLE_TYPE(BRID, Bridge));
+                typeIDTypeMap.insert(HANDLE_TYPE(BRID, BridgeConstructionElement));
+                typeIDTypeMap.insert(HANDLE_TYPE(BRID, BridgeInstallation));
+                typeIDTypeMap.insert(HANDLE_TYPE(BRID, BridgePart));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, WallSurface));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, RoofSurface));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, GroundSurface));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, ClosureSurface));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, FloorSurface));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, InteriorWallSurface));
+                typeIDTypeMap.insert(HANDLE_TYPE(BLDG, CeilingSurface));
+
+                typeIDTypeMapInitialized = true;
+            }
+        }
     }
 
     void CityObjectElementParser::initializeAttributesSet()
     {
-        attributesSet.insert(HANDLE_ATTR(CORE, CreationDate));
-        attributesSet.insert(HANDLE_ATTR(CORE, TerminationDate));
-        attributesSet.insert(HANDLE_ATTR(BLDG, Type));
-        attributesSet.insert(HANDLE_ATTR(BLDG, Class));
-        attributesSet.insert(HANDLE_ATTR(BLDG, Function));
-        attributesSet.insert(HANDLE_ATTR(BLDG, Usage));
-        attributesSet.insert(HANDLE_ATTR(BLDG, YearOfConstruction));
-        attributesSet.insert(HANDLE_ATTR(BLDG, YearOfDemolition));
-        attributesSet.insert(HANDLE_ATTR(BLDG, StoreyHeightsAboveGround));
-        attributesSet.insert(HANDLE_ATTR(BLDG, StoreysBelowGround));
-        attributesSet.insert(HANDLE_ATTR(BLDG, MeasuredHeight));
-        attributesSet.insert(HANDLE_ATTR(BLDG, Address));
-        attributesSet.insert(HANDLE_ATTR(BLDG, RoofType));
-        attributesSet.insert(HANDLE_ATTR(XAL, XalAddress));
-        attributesSet.insert(HANDLE_ATTR(XAL, Administrativearea));
-        attributesSet.insert(HANDLE_ATTR(XAL, Country));
-        attributesSet.insert(HANDLE_ATTR(XAL, CountryName));
-        attributesSet.insert(HANDLE_ATTR(XAL, Code));
-        attributesSet.insert(HANDLE_ATTR(XAL, Street));
-        attributesSet.insert(HANDLE_ATTR(XAL, PostalCode));
-        attributesSet.insert(HANDLE_ATTR(XAL, City));
-        attributesSet.insert(HANDLE_ATTR(XAL, LocalityName));
-        attributesSet.insert(HANDLE_ATTR(XAL, Thoroughfare));
-        attributesSet.insert(HANDLE_ATTR(XAL, ThoroughfareNumber));
-        attributesSet.insert(HANDLE_ATTR(XAL, ThoroughfareName));
-        attributesSet.insert(HANDLE_ATTR(XAL, Locality));
-        attributesSet.insert(HANDLE_ATTR(XAL, AddressDetails));
-        attributesSet.insert(HANDLE_ATTR(XAL, DependentLocalityName));
-
-        attributesSetInitialized = true;
+        // double-checked lock
+        if (!attributesSetInitialized) {
+            std::lock_guard<std::mutex> lock(CityObjectElementParser::initializedAttributeSetMutex);
```

Comment:
shouldn't this be moved out of the code block so that the if block cannot be entered twice in different threads